### PR TITLE
TRS import via URL

### PR DESCRIPTION
**Pull request by fbacall**
Related issue: #14636

Apologies if @dannon is already working on this, feel free to close if so.

Provides an alternate means for importing Galaxy workflows via the TRS API using a URL, without the Galaxy instance having prior configuration of the service providing it.

I mostly copied the behaviour of the existing TRS import, but am not a Python programmer so may have broken conventions here and there. I also wasn't sure if I implemented the Vue component in a sensible way.

UI-wise it's an extra box under the current TRS import form:
image

although I expect this would mostly be used by external services providing the TRS URL directly to Galaxy to import.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Visit https://localhost:8080/workflows/trs_import and fill in the TRS URL field with a valid TRS URL to a Galaxy workflow (e.g. https://workflowhub.eu/ga4gh/trs/v2/tools/109/versions/5)
  3. Or to import automatically, visit e.g. https://localhost:8080/workflows/trs_import?trs_url=https://dockstore.org/api/ga4gh/trs/v2/tools/%23workflow%2Fgithub.com%2Fjmchilton%2Fgalaxy-workflow-dockstore-example-1%2Fmycoolworkflow/versions/master

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by dannon** [comment]
> Apologies if @dannon is already working on this, feel free to close if so.

Absolutely no worries here -- it was on my list but now doesn't have to be ;)

File: test/unit/workflows/test_trs_proxy.py [hunk 1]
```diff
@@ -19,50 +19,132 @@
...
+
+    ignore_extra = proxy.match_url(
+        "https://workflowhub.eu/ga4gh/trs/v2/tools/344/versions/1/CWL/descriptor/ro-crate-metadata.json")
+    assert ignore_extra
+    assert ignore_extra["trs_base_url"] == "https://workflowhub.eu"
+    assert ignore_extra["tool_id"] == "344"
+    assert ignore_extra["version_id"] == "1"
+
+    invalid = proxy.match_url("https://workflowhub.eu/workflows/1")
+    assert invalid == None
```
Comment (nsoranzo) [inline test/unit/workflows/test_trs_proxy.py]:
```suggestion
    assert invalid is None
```

File: test/unit/workflows/test_trs_proxy.py [hunk 2]
```diff
@@ -19,50 +19,132 @@
...
+
+    missing_version = proxy.match_url("https://workflowhub.eu/ga4gh/trs/v2/tools/344")
+    assert missing_version == None
```
Comment (nsoranzo) [inline test/unit/workflows/test_trs_proxy.py]:
```suggestion
    assert missing_version is None
```

File: test/unit/workflows/test_trs_proxy.py [hunk 3]
```diff
@@ -19,50 +19,132 @@
...
+
+    blank = proxy.match_url("")
+    assert blank == None
```
Comment (nsoranzo) [inline test/unit/workflows/test_trs_proxy.py]:
```suggestion
    assert blank is None
```

File: test/unit/workflows/test_trs_proxy.py [hunk 4]
```diff
@@ -19,50 +19,132 @@
...
+
+    not_url = proxy.match_url("1234")
+    assert not_url == None
```
Comment (nsoranzo) [inline test/unit/workflows/test_trs_proxy.py]:
```suggestion
    assert not_url is None
```

**Review by nsoranzo — COMMENTED** [review]
Thanks @fbacall !

If you run `make format` from the root directory of your `galaxy` clone, it should fix all other linting issues reported in https://github.com/galaxyproject/galaxy/actions/runs/3394635717/jobs/5643447167 .

**Comment by nsoranzo** [comment]
And `make client-format` to fix the Javascript linting, thanks!

**Comment by fbacall** [comment]
Looking at the selenium tests now, somehow I missed them when looking at the existing TRS code.

File: lib/galaxy/managers/workflows.py [hunk 5]
```diff
@@ -720,6 +720,7 @@ def _workflow_from_raw_description(
                     source_metadata["trs_tool_id"] = workflow_state_resolution_options.trs_tool_id
                     source_metadata["trs_version_id"] = workflow_state_resolution_options.trs_version_id
                     source_metadata["trs_server"] = workflow_state_resolution_options.trs_server
+                    source_metadata["trs_url"] = workflow_state_resolution_options.trs_url
```
Comment (fbacall) [inline lib/galaxy/managers/workflows.py:735]: I wasn't sure what this `source_metadata` is used for, and if there are any consequences to a workflow not having a `trs_server` defined if it was imported directly from a URL.

File: lib/galaxy/webapps/galaxy/api/workflows.py [hunk 6]
```diff
@@ -340,11 +340,26 @@ def create(self, trans: GalaxyWebTransaction, payload=None, **kwd):
                     payload["workflow"] = workflow_src
                     return self.__api_import_new_workflow(trans, payload, **kwd)
                 elif archive_source == "trs_tool":
-                    trs_server = payload.get("trs_server")
+                    server = None
+                    trs_tool_id = None
+                    trs_version_id = None
+                    import_source = None
+                    if "trs_url" in payload:
+                        parts = self.app.trs_proxy.match_url(payload.get("trs_url"))
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/workflows.py]:
```suggestion
                        parts = self.app.trs_proxy.match_url(payload["trs_url"])
```

File: lib/galaxy/webapps/galaxy/api/trs_consumer.py [hunk 7]
```diff
@@ -21,12 +21,12 @@ def get_servers(self, trans, *args, **kwd):
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/api/trs_consumer.py:21]: Just noticed that this (`/api/trs_consume/`) and `/api/trs_search` don't have any API test in `lib/galaxy_test/api/` . Not the fault of this PR though, and I guess the endpoints are exercised by the Selenium tests.

File: lib/galaxy/managers/workflows.py (hunk 5 above)
Comment (jmchilton) [reply lib/galaxy/managers/workflows.py:735]: You're capturing all the relevant information and the tests below are really nice. I think the goal was to be able to upgrade the workflow after upgrades and to provide linkage information in the UI - I think what you're capturing here is sufficient to do either.

**Comment by fbacall** [comment]
Awesome, thanks for the feedback @jmchilton and @nsoranzo !

**Comment by jmchilton** [comment]
Thanks for your patience and thanks for the excellent PR!

**Pull request merged**